File: streampark-flink/streampark-flink-kubernetes/src/main/java/org/apache/streampark/flink/kubernetes/watcher/FlinkCheckpointWatcher.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes.watcher;

import org.apache.streampark.flink.kubernetes.ChangeEventBus;
import org.apache.streampark.flink.kubernetes.FlinkK8sWatchController;
import org.apache.streampark.flink.kubernetes.KubernetesRetriever;
import org.apache.streampark.flink.kubernetes.MetricWatcherConfig;
import org.apache.streampark.flink.kubernetes.event.FlinkJobCheckpointChangeEvent;
import org.apache.streampark.flink.kubernetes.model.CheckpointCV;
import org.apache.streampark.flink.kubernetes.model.ClusterKey;
import org.apache.streampark.flink.kubernetes.model.TrackId;
import org.apache.streampark.flink.kubernetes.rest.CheckpointInfo;
import org.apache.streampark.flink.kubernetes.rest.FlinkCheckpointResponse;

import org.apache.hc.client5.http.fluent.Request;

import javax.annotation.concurrent.ThreadSafe;

import java.nio.charset.StandardCharsets;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.stream.Collectors;

@ThreadSafe
public class FlinkCheckpointWatcher extends FlinkWatcher {

    private final MetricWatcherConfig conf;
    private final FlinkK8sWatchController watchController;
    private final ChangeEventBus eventBus;

    private ScheduledFuture<?> timerSchedule;

    public FlinkCheckpointWatcher(
                                  MetricWatcherConfig conf,
                                  FlinkK8sWatchController watchController,
                                  ChangeEventBus eventBus) {
        this.conf = conf;
        this.watchController = watchController;
        this.eventBus = eventBus;
    }

    @Override
    protected void doStart() {
        timerSchedule =
            watchExecutor.scheduleAtFixedRate(
                this::doWatch, 0, conf.requestIntervalSec(), TimeUnit.SECONDS);
        logInfo("[flink-k8s] FlinkCheckpointWatcher started.");
    }

    @Override
    protected void doStop() {
        if (timerSchedule != null && !timerSchedule.isCancelled()) {
            timerSchedule.cancel(true);
        }
        logInfo("[flink-k8s] FlinkCheckpointWatcher stopped.");
    }

    @Override
    protected void doClose() {
        if (timerSchedule != null && !timerSchedule.isCancelled()) {
            timerSchedule.cancel(true);
        }
        logInfo("[flink-k8s] FlinkCheckpointWatcher closed.");
    }

    @Override
    public void doWatch() {
        Set<TrackId> trackIds;
        try {
            trackIds = watchController.getActiveWatchingIds();
            if (trackIds.isEmpty()) {
                return;
            }
        } catch (Exception e) {
            return;
        }

        Set<CompletableFuture<Optional<CheckpointCV>>> futures =
            trackIds.stream()
                .map(
                    id -> CompletableFuture
                        .supplyAsync(() -> collect(id), watchExecutor)
                        .whenComplete(
                            (cpOpt, error) -> {
                                if (error == null) {
                                    cpOpt.ifPresent(
                                        cp -> eventBus.postAsync(
                                            new FlinkJobCheckpointChangeEvent(id, cp)));
                                }
                            }))
                .collect(Collectors.toSet());

        try {
            CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
                .get(conf.requestTimeoutSec(), TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            logError(
                "[FlinkCheckpointWatcher] tracking flink-job checkpoint on kubernetes mode interrupted,"
                    + " limitSeconds="
                    + conf.requestTimeoutSec()
                    + ", trackingClusterKeys="
                    + trackIds.stream().map(Object::toString).collect(Collectors.joining(",")));
        } catch (ExecutionException | TimeoutException e) {
            logError(
                "[FlinkCheckpointWatcher] tracking flink-job checkpoint on kubernetes mode timeout,"
                    + " limitSeconds="
                    + conf.requestTimeoutSec()
                    + ", trackingClusterKeys="
                    + trackIds.stream().map(Object::toString).collect(Collectors.joining(",")));
        }
    }

    public Optional<CheckpointCV> collect(TrackId trackId) {
        if (trackId.jobId() == null) {
            return Optional.empty();
        }
        Optional<String> flinkJmRestUrl =
            watchController
                .getClusterRestUrl(ClusterKey.of(trackId))
                .filter(url -> !url.isEmpty());
        if (!flinkJmRestUrl.isPresent()) {
            return Optional.empty();
        }
        try {
            String json =
                Request.get(flinkJmRestUrl.get() + "/jobs/" + trackId.jobId() + "/checkpoints")
                    .connectTimeout(KubernetesRetriever.FLINK_REST_AWAIT_TIMEOUT_SEC)
                    .responseTimeout(KubernetesRetriever.FLINK_CLIENT_TIMEOUT_SEC)
                    .execute()
                    .returnContent()
                    .asString(StandardCharsets.UTF_8);
            Optional<CheckpointInfo> checkpoint = FlinkCheckpointResponse.parseCompleted(json);
            if (!checkpoint.isPresent()) {
                return Optional.empty();
            }
            CheckpointInfo cp = checkpoint.get();
            return Optional.of(
                CheckpointCV.builder()
                    .id(cp.id())
                    .externalPath(cp.externalPath())
                    .isSavepoint(cp.isSavepoint())
                    .checkpointType(cp.checkpointType())
                    .status(cp.status())
                    .triggerTimestamp(cp.triggerTimestamp())
                    .build());
        } catch (Exception e) {
            return Optional.empty();
        }
    }
}

File: streampark-flink/streampark-flink-kubernetes/src/main/scala/org/apache/streampark/flink/kubernetes/watcher/FlinkCheckpointWatcher.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.kubernetes.watcher

import org.apache.streampark.common.util.Logger
import org.apache.streampark.flink.kubernetes.{ChangeEventBus, FlinkK8sWatchController, KubernetesRetriever, MetricWatcherConfig}
import org.apache.streampark.flink.kubernetes.event.FlinkJobCheckpointChangeEvent
import org.apache.streampark.flink.kubernetes.model.{CheckpointCV, ClusterKey, TrackId}

import org.apache.hc.client5.http.fluent.Request
import org.json4s.{DefaultFormats, JNull}
import org.json4s.JsonAST.JNothing
import org.json4s.jackson.JsonMethods.parse

import javax.annotation.concurrent.ThreadSafe

import java.nio.charset.StandardCharsets
import java.util.concurrent.{ScheduledFuture, TimeUnit}

import scala.concurrent.{Await, ExecutionContext, ExecutionContextExecutorService, Future}
import scala.concurrent.duration.DurationLong
import scala.language.postfixOps
import scala.util.{Failure, Success, Try}

@ThreadSafe
class FlinkCheckpointWatcher(conf: MetricWatcherConfig = MetricWatcherConfig.defaultConf)(
    implicit val watchController: FlinkK8sWatchController,
    implicit val eventBus: ChangeEventBus)
  extends Logger
  with FlinkWatcher {

  implicit private val trackTaskExecutor: ExecutionContextExecutorService =
    ExecutionContext.fromExecutorService(watchExecutor)

  private var timerSchedule: ScheduledFuture[_] = _

  /** start watcher process */
  override def doStart(): Unit = {
    timerSchedule = watchExecutor.scheduleAtFixedRate(
      () => doWatch(),
      0,
      conf.requestIntervalSec,
      TimeUnit.SECONDS)
    logInfo("[flink-k8s] FlinkCheckpointWatcher started.")
  }

  /** stop watcher process */
  override def doStop(): Unit = {
    if (!timerSchedule.isCancelled) {
      timerSchedule.cancel(true)
    }
    logInfo("[flink-k8s] FlinkCheckpointWatcher stopped.")
  }

  /** closes resource, relinquishing any underlying resources. */
  override def doClose(): Unit = {
    if (Option(timerSchedule).isDefined && !timerSchedule.isCancelled) {
      timerSchedule.cancel(true)
    }
    logInfo("[flink-k8s] FlinkCheckpointWatcher closed.")
  }

  /** single flink metrics tracking task */
  override def doWatch(): Unit = {
    // get all legal tracking cluster key
    val trackIds: Set[TrackId] = Try(watchController.getActiveWatchingIds())
      .filter(_.nonEmpty)
      .getOrElse(return
      )
    // retrieve flink metrics in thread pool
    val futures: Set[Future[Option[CheckpointCV]]] =
      trackIds.map(
        id => {
          val future = Future(collect(id))
          future.onComplete(_.getOrElse(None) match {
            case Some(cp) => eventBus.postAsync(FlinkJobCheckpointChangeEvent(id, cp))
            case _ =>
          })
          future
        })

    // blocking until all future are completed or timeout is reached
    Try(Await.ready(Future.sequence(futures), conf.requestTimeoutSec seconds)).failed.map {
      _ =>
        logError(
          s"[FlinkCheckpointWatcher] tracking flink-job checkpoint on kubernetes mode timeout," +
            s" limitSeconds=${conf.requestTimeoutSec}," +
            s" trackingClusterKeys=${trackIds.mkString(",")}")
    }
  }

  /**
   * Collect flink-job checkpoint from kubernetes-native cluster. Returns None when the
   * flink-cluster-client request fails (or in case of the relevant flink rest api require failure).
   */
  def collect(trackId: TrackId): Option[CheckpointCV] = {
    if (trackId.jobId != null) {
      val flinkJmRestUrl = watchController
        .getClusterRestUrl(ClusterKey.of(trackId))
        .filter(_.nonEmpty)
        .getOrElse(return None)
      // call flink rest overview api
      val checkpoint: Checkpoint = Try(
        Checkpoint.as(
          Request
            .get(s"$flinkJmRestUrl/jobs/${trackId.jobId}/checkpoints")
            .connectTimeout(KubernetesRetriever.FLINK_REST_AWAIT_TIMEOUT_SEC)
            .responseTimeout(KubernetesRetriever.FLINK_CLIENT_TIMEOUT_SEC)
            .execute
            .returnContent
            .asString(StandardCharsets.UTF_8)) match {
          case Some(c) => c
          case _ => return None
        }).getOrElse(return None)

      val checkpointCV = CheckpointCV(
        id = checkpoint.id,
        externalPath = checkpoint.externalPath,
        isSavepoint = checkpoint.isSavepoint,
        checkpointType = checkpoint.checkpointType,
        status = checkpoint.status,
        triggerTimestamp = checkpoint.triggerTimestamp
      )
      Some(checkpointCV)
    } else None
  }

}

private[kubernetes] case class Checkpoint(
    id: Long,
    status: String,
    externalPath: String,
    isSavepoint: Boolean,
    checkpointType: String,
    triggerTimestamp: Long)

object Checkpoint {

  @transient
  implicit lazy val formats: DefaultFormats.type = org.json4s.DefaultFormats

  def as(json: String): Option[Checkpoint] = {
    Try(parse(json)) match {
      case Success(ok) =>
        val completed = ok \ "latest" \ "completed"
        completed match {
          case JNull | JNothing => None
          case _ =>
            val cp = Checkpoint(
              id = (completed \ "id").extractOpt[Long].getOrElse(0L),
              status = (completed \ "status").extractOpt[String].orNull,
              externalPath = (completed \ "external_path").extractOpt[String].orNull,
              isSavepoint = (completed \ "is_savepoint").extractOpt[Boolean].getOrElse(false),
              checkpointType = (completed \ "checkpoint_type").extractOpt[String].orNull,
              triggerTimestamp = (completed \ "trigger_timestamp").extractOpt[Long].getOrElse(0L)
            )
            Some(cp)
        }
      case Failure(_) => None
    }
  }

}
